setup.sh

```diff
@@ -0,0 +1,134 @@
+#!/bin/bash
+
+# Usage: setup.sh [--help]
+#
+# Environment variables:
+#   VENV: use the specified Python virtual environment. If empty, don't use a
+#     virtual environment and use python/pip from the caller's $PATH. Defaults
+#     to $VIRTUAL_ENV or "$BASEDIR/ocp-venv".
+#   VENV_PYTHON: the python interpreter to use to create the virtual
+#     environment. This only matters, while creating $VENV. If the virtual
+#     environment already exists, or if we are called with VENV="" then this has
+#     no effect. Aside creating the $VENT, we use python/pip from the $PATH/$VENV.
+#   VERBOSE: if set to non-empty, don't hide output of commands.
+#
+# - If you use VENV, you will need to `source "$VENT/bin/activate" before using cda.py.
+# - Sets up bash completion for `oc`. You may want to start a new bash
+#   afterwards or source /etc/bash_completion.d/oc_bash_completion.
+# - If ~/.ssh/id_ed25519 does not exist, a new key gets generated.
+# - The script changes into the top dir of cluster-deployment-automation. All path
+#   names ($VENV) are relative to that path.
+
+set -e
+
+die() {
+    printf '%s\n' "$*"
+    exit 1
+}
+
+usage() {
+    sed -n '3,/^$/ s/^#\( \(.*\)\)\?$/\2/p' "$0"
+}
+
+# Parse command line arguments. None are supported, except --help. We always
+# print the usage.
+if [ "$#" -ge 1 ] ; then
+    usage
+    for c ; do
+        if [ "$c" != "-h" -a "$c" != "--help" -a "$c" != "help" ] ; then
+            printf '\n%s\n' "ERROR: No command line arguments supported. Set environment variables."
+            exit 2
+        fi
+    done
+    exit 0
+fi
+
+do_cmd() {
+    local outfile
+    local rc
+
+    if [ -z "$BANNER" ] ; then
+        local BANNER="$(printf ' %q' "$@")"
+        BANNER="${BANNER:1}"
+    fi
+    printf '%s\n' "RUN:  $BANNER"
+
+    rc=0
+    if [ -z "$VERBOSE" ] ; then
+        outfile="$(mktemp --tmpdir cda-setup.XXXXXXX)"
+        "$@" &> "$outfile" || rc="$?"
+    else
+        outfile=
+        "$@" || rc="$?"
+    fi
+
+    if [ "$rc" -eq 0 ] ; then
+        return 0
+    fi
+
+    if [ -n "$outfile" ] ; then
+        cat "$outfile"
+        rm -f "$outfile"
+    fi
+    die "RUN:  Failure to run command \`$(printf '%q ' "$@")\`"
+}
+
+BASEDIR="$(dirname "$0")"
+cd "$BASEDIR" || die "Could not change into base directory for \"$0\""
+test -f "$BASEDIR/cda.py" || die "Base directory for \"$0\" does not look like cluster-deployment-automation source tree"
+
+[ "$EUID" = 0 ] || die "Must run as root user"
+
+_USE_VENV=0
+if [ -z "${VENV+x}" -a -n "$VIRTUAL_ENV" ] ; then
```

Comment:
The moment we start adding "real logic" in a shell script is the moment we want to start thinking about moving it into a python script.